File: source/libs/data/objects/components/Transform.cpp
#include "Transform.h"
#include "../Object.h"
#include <nlohmann/json.hpp>
#include <Protocol.h>

Transform::Transform()
  : Component(ComponentType::transform)
{
  loadVariable(m_position);
  loadVariable(m_scale);
  loadVariable(m_rotation);
}

Transform::Transform(const glm::vec3& position, const glm::vec3& scale, const glm::vec3& rotation)
  : Component(ComponentType::transform),
    m_position(position),
    m_scale(scale),
    m_rotation(rotation)
{
  loadVariable(m_position);
  loadVariable(m_scale);
  loadVariable(m_rotation);
}

uint8_t Transform::getUpdateID() const
{
  return m_updateID;
}

glm::vec3 Transform::getPosition() const
{
  if (m_owner->getParent())
  {
    if (const auto& parentTransform = m_owner->getParent()->getComponent<Transform>(ComponentType::transform))
    {
      return parentTransform->getPosition() + m_position.get();
    }
  }

  return m_position.get();
}

glm::vec3 Transform::getScale() const
{
  if (m_owner->getParent())
  {
    if (const auto& parentTransform = m_owner->getParent()->getComponent<Transform>(ComponentType::transform))
    {
      return parentTransform->getScale() * m_scale.get();
    }
  }

  return m_scale.get();
}

glm::vec3 Transform::getRotation() const
{
  if (m_owner->getParent())
  {
    if (const auto& parentTransform = m_owner->getParent()->getComponent<Transform>(ComponentType::transform))
    {
      return parentTransform->getRotation() + m_rotation.get();
    }
  }

  return m_rotation.get();
}

glm::vec3 Transform::getLocalPosition() const
{
  return m_position.get();
}

glm::vec3 Transform::getLocalScale() const
{
  return m_scale.get();
}

glm::vec3 Transform::getLocalRotation() const
{
  return m_rotation.get();
}

void Transform::setPosition(const glm::vec3 position)
{
  m_position.set(position);
  ++m_updateID;
}

void Transform::setScale(const glm::vec3 scale)
{
  m_scale.set(scale);
  ++m_updateID;
}

void Transform::setRotation(const glm::vec3 rotation)
{
  m_rotation.set(rotation);
  ++m_updateID;
}

void Transform::move(const glm::vec3& direction)
{
  m_position.value() += direction;
  ++m_updateID;
}

nlohmann::json Transform::serialize()
{
  const auto position = m_position.getInitialValue();
  const auto rotation = m_rotation.getInitialValue();
  const auto scale = m_scale.getInitialValue();

  const nlohmann::json data = {
    { "type", "Transform" },
    { "position", { position.x, position.y, position.z } },
    { "rotation", { rotation.x, rotation.y, rotation.z } },
    { "scale", { scale.x, scale.y, scale.z } }
  };

  return data;
}

void Transform::loadFromJSON(const nlohmann::json& componentData)
{
  const auto& position = componentData.at("position");
  const auto& rotation = componentData.at("rotation");
  const auto& scale = componentData.at("scale");

  m_position.set(glm::vec3(position.at(0), position.at(1), position.at(2)));
  m_rotation.set(glm::vec3(rotation.at(0), rotation.at(1), rotation.at(2)));
  m_scale.set(glm::vec3(scale.at(0), scale.at(1), scale.at(2)));
}

void Transform::pack(net::Message& message) const
{
  message.write(ComponentType::transform);

  message.write(m_position.get());
  message.write(m_scale.get());
  message.write(m_rotation.get());
}

void Transform::unpack(net::MessageReader& messageReader)
{
  m_position.set(messageReader.read<glm::vec3>());
  m_scale.set(messageReader.read<glm::vec3>());
  m_rotation.set(messageReader.read<glm::vec3>());
}

File: source/apps/webClient/src/data/objects/components/Transform.ts
// Port of source/libs/data/objects/components/Transform.{h,cpp}.
//
// The world-space getters compose with the parent the same (non-TRS) way upstream does: position and
// rotation ADD, scale MULTIPLIES. That is not a real transform hierarchy - a rotated parent does not
// orbit its children - but the server replicates LOCAL transforms and composes identically, so this
// must match exactly or every parented object lands somewhere the server didn't put it.
// (ROADMAP.md D5 tracks fixing it upstream; when it lands, this changes with it.)

import type { MessageReader, Vec3 } from "../../../net/Protocol";
import { Component, ComponentType, vec3FromJSON } from "./Component";

export class Transform extends Component {
  private position: Vec3 = [0, 0, 0];
  private scale: Vec3 = [0, 0, 0];
  private rotation: Vec3 = [0, 0, 0];

  constructor() {
    super(ComponentType.transform);
  }

  getLocalPosition(): Vec3 {
    return this.position;
  }

  getLocalScale(): Vec3 {
    return this.scale;
  }

  getLocalRotation(): Vec3 {
    return this.rotation;
  }

  setLocalPosition(position: Vec3): void {
    this.position = position;
  }

  setLocalScale(scale: Vec3): void {
    this.scale = scale;
  }

  setLocalRotation(rotation: Vec3): void {
    this.rotation = rotation;
  }

  getPosition(): Vec3 {
    const parent = this.parentTransform();
    if (!parent) {
      return this.position;
    }

    const base = parent.getPosition();
    return [base[0] + this.position[0], base[1] + this.position[1], base[2] + this.position[2]];
  }

  getScale(): Vec3 {
    const parent = this.parentTransform();
    if (!parent) {
      return this.scale;
    }

    const base = parent.getScale();
    return [base[0] * this.scale[0], base[1] * this.scale[1], base[2] * this.scale[2]];
  }

  getRotation(): Vec3 {
    const parent = this.parentTransform();
    if (!parent) {
      return this.rotation;
    }

    const base = parent.getRotation();
    return [base[0] + this.rotation[0], base[1] + this.rotation[1], base[2] + this.rotation[2]];
  }

  // Packed order is position, scale, rotation - NOT the position/rotation/scale order the JSON blob
  // and the state delta both use.
  unpack(reader: MessageReader): void {
    this.position = reader.readVec3();
    this.scale = reader.readVec3();
    this.rotation = reader.readVec3();
  }

  loadFromJSON(data: Record<string, unknown>): void {
    this.position = vec3FromJSON(data.position, this.position);
    this.rotation = vec3FromJSON(data.rotation, this.rotation);
    this.scale = vec3FromJSON(data.scale, this.scale);
  }

  private parentTransform(): Transform | null {
    const parent = this.owner?.getParent();
    return parent?.getComponent<Transform>(ComponentType.transform) ?? null;
  }
}
